pkg/image/importer/credentials.go

```diff
@@ -176,12 +183,17 @@ func basicCredentialsFromKeyring(keyring credentialprovider.DockerKeyring, targe
 		// do a special case check for docker.io to match historical lookups when we respond to a challenge
 		if value == "auth.docker.io/token" {
 			glog.V(5).Infof("Being asked for %s, trying %s for legacy behavior", target, "index.docker.io/v1")
-			return basicCredentialsFromKeyring(keyring, &url.URL{Host: "index.docker.io", Path: "/v1"})
+			return basicCredentialsFromKeyring(keyring, &url.URL{Host: "index.docker.io", Path: "/v1"}, nil)
 		}
 		// docker 1.9 saves 'docker.io' in config in f23, see https://bugzilla.redhat.com/show_bug.cgi?id=1309739
 		if value == "index.docker.io" {
 			glog.V(5).Infof("Being asked for %s, trying %s for legacy behavior", target, "docker.io")
-			return basicCredentialsFromKeyring(keyring, &url.URL{Host: "docker.io"})
+			return basicCredentialsFromKeyring(keyring, &url.URL{Host: "docker.io"}, nil)
+		}
+		// try origin url, if it's different than target
+		if origin != nil && origin.String() != target.String() {
```

Comment:
There is no guarantee that `origin` really corresponds to the `target` realm. In docker-registry, single context can be used for different registries, each having possibly different auth server. The association `origin <-> target` needs to be verified first. Otherwise, credentials belonging to one server may be sent to another.

The verification may be simple like extracting auth realm from a simple http GET request to the `origin`:

```
curl -I https://index.docker.io/v2/
HTTP/1.1 401 Unauthorized
Content-Type: application/json; charset=utf-8
Docker-Distribution-Api-Version: registry/2.0
Www-Authenticate: Bearer realm="https://auth.docker.io/token",service="registry.docker.io"
Date: Thu, 29 Jun 2017 11:53:49 GMT
```

Thread-safety needs to be ensured though. I would welcome the verification being done lazily. However, I'm not sure about polling remote registries in this function which is supposed to be fast.

@liggitt this is your domain, do you have better idea?